pkg/codegen/go/gen_program_expressions.go

```diff
@@ -375,20 +496,84 @@ func (g *generator) argumentTypeName(expr model.Expression, destType model.Type,
 		}
 	}
 
-	// TODO support rest of types
 	switch destType := destType.(type) {
 	case *model.OpaqueType:
 		switch destType {
+		case model.IntType:
+			if isInput {
+				return "pulumi.Int"
+			}
+			return "int"
 		case model.NumberType:
+			if isInput {
+				return "pulumi.Float64"
+			}
 			return "float64"
+		case model.StringType:
+			if isInput {
+				return "pulumi.String"
+			}
+			return "string"
+		case model.BoolType:
+			if isInput {
+				return "pulumi.Bool"
+			}
+			return "bool"
 		default:
 			return destType.Name
 		}
-	// TODO could probably improve these types by inspecting kv types
-	case *model.ObjectType, *model.MapType:
+	case *model.ObjectType:
+		return "map[string]interface{}"
+	case *model.MapType:
+		valType := argumentTypeName(nil, destType.ElementType, isInput)
+		return fmt.Sprintf("map[string]%s", valType)
+	case *model.ListType:
+		argTypeName := argumentTypeName(nil, destType.ElementType, isInput)
+		if strings.HasPrefix(argTypeName, "pulumi.") {
+			return fmt.Sprintf("%sArray", argTypeName)
+		}
+		return fmt.Sprintf("[]%s", argTypeName)
```

Comment:
I think these will need to handle input-typed elements and generate inputty names (e.g. `FooMap`, `FooArray`, etc.). Not something to fix now, but maybe file an issue for follow-up.

Reply:
https://github.com/pulumi/pulumi/issues/4763